csharp/src/Microsoft.ML.OnnxRuntime/Training/CheckpointState.shared.cs

```diff
@@ -47,19 +41,126 @@ public CheckpointState(string checkpointPath)
+            {
+                IntPtr memPtr;
+                unsafe
+                {
+                    memPtr = (IntPtr)memHandle.Pointer;
+                }
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtAddProperty(handle, propertyNameUtf8, (long)PropertyType.Float, memPtr));
+            }
+        }
+
+        /// <summary>
+        /// Adds the given string property to the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being added.</param>
+        /// <param name="propertyValue">Property value associated with the given name.</param>
+        /// </summary>
+        public void AddProperty(string propertyName, string propertyValue)
+        {
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            var propertyValueUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyValue);
+
+            IntPtr unmanagedPointer = Marshal.AllocHGlobal(propertyValueUtf8.Length);
+            try
+            {
+                Marshal.Copy(propertyValueUtf8, 0, unmanagedPointer, propertyValueUtf8.Length);
+                NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtAddProperty(handle, propertyNameUtf8, (long)PropertyType.String, unmanagedPointer));
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(unmanagedPointer);
+            }
+        }
+
+        /// <summary>
+        /// Gets the property value associated with the given name from the checkpoint state.
+        /// <param name="propertyName">Unique name of the property being retrieved.</param>
+        /// </summary>
+        public object GetProperty(string propertyName)
+        {
+            var propertyNameUtf8 = NativeOnnxValueHelper.StringToZeroTerminatedUtf8(propertyName);
+            var allocator = OrtAllocator.DefaultInstance;
+            IntPtr propertyValue = IntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeTrainingMethods.OrtGetProperty(handle, propertyNameUtf8, allocator.Pointer, out long propertyType, out propertyValue));
+
+            if (propertyType == (long)PropertyType.Int)
+            {
+                var longPropertyValue = Marshal.ReadInt64(propertyValue);
+                allocator.FreeMemory(propertyValue);
+                return longPropertyValue;
+            }
+            else if (propertyType == (long)PropertyType.Float)
+            {
+                float[] value = new float[1];
+                Marshal.Copy(propertyValue, value, 0, 1);
+                allocator.FreeMemory(propertyValue);
+                return value[0];
+            }
+            else if (propertyType == (long)PropertyType.String)
+            {
+                return NativeOnnxValueHelper.StringFromNativeUtf8(propertyValue, allocator);
+            }
+
+            throw new ArgumentException("Expected the property type to be one of long, float or string. Unknown type retrieved.");
```

Comment:
nit: include the type value that was received in the error message so it's more easily debugged.

Reply:
Actually, control should never reach this exception. The C code only knows these 3 types for property types (long, float and string). Any other type is unexpected from the C code.

Not sure if you have any recommendations in such a case. Maybe I can add an assert to indicate that control should never reach here?